### ISSUE TITLE
load-from-h2 not working because of circular references between RawColumns

### ISSUE THREAD

**Issue by snarfed** [metabase/metabase#2632]
hey guys! i'm working on migrating from H2 to Postgres again, and i'm close, but hitting a new error. it looks like you maybe need to tweak the table ordering so that foreign key constraints are satisfied? or something?

i upgraded to 0.17.1 successfully today, then later stopped metabase, backed up `metabase.db.h2.db`, and restarted. when i set up and run `load-from-h2`, i get:

```
$ java -jar ~/metabase.jar load-from-h2
05-20 00:31:02 INFO metabase.db :: Verifying Database Connection ...
05-20 00:31:02 DEBUG metabase.driver :: Registered driver :postgres 🚚
05-20 00:31:02 INFO metabase.db :: Verify Database Connection ... ✅
05-20 00:32:02 INFO metabase.db :: Database Migrations Current ... ✅
05-20 00:32:02 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
05-20 00:32:03 INFO db.migrations :: Running data migration 'set-card-database-and-table-ids'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'set-mongodb-databases-ssl-false'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'set-default-schemas'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'set-admin-email'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'remove-database-sync-activity-entries'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'remove-duplicate-fk-entries'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'update-dashboards-to-new-grid'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'migrate-field-visibility-type'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'fix-dashboard-cards-without-positions'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'migrate-fk-metadata'...
05-20 00:32:03 INFO db.migrations :: Running data migration 'create-raw-tables'...
05-20 00:32:03 INFO db.migrations :: Finished running data migrations.
Transfering 4 instances of Database....[OK]
Transfering 220 instances of RawTable....[OK]
Transfering 1672 instances of RawColumn....org.postgresql.util.PSQLException: ERROR: insert or update on table "raw_column" violates foreign key constraint "fk_rawcolumn_fktarget_ref_rawcolumn"
  Detail: Key (fk_target_column_id)=(766) is not present in table "raw_column".
...
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:419)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeUpdate(AbstractJdbc2Statement.java:365)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:105)
...
	at metabase.cmd.load_from_h2$load_from_h2.invokeStatic(load_from_h2.clj:80)
	at metabase.cmd.load_from_h2$load_from_h2.invoke(load_from_h2.clj:70)
...
Command failed with exception: ERROR: insert or update on table "raw_column" violates foreign key constraint "fk_rawcolumn_fktarget_ref_rawcolumn"
  Detail: Key (fk_target_column_id)=(766) is not present in table "raw_column".
```

i get the same error when i run it locally, on a mac, with the same jar and H2 db into postgres.

thanks in advance!

*  My server's OS is Ubuntu 14.04.
*  I'm running Metabase 0.17.1 on AWS EC2, from the JAR.
*  My database is H2, trying to port to Postgres 9.3.10 on RDS.

**Comment by camsaul** [comment]
This will be tricky to fix... the new `RawColumn` model from #2372 has circular references to other `RawColumns` so we'll have to rework the code to work around that.

**Comment by snarfed** [comment]
if you load all of the circularly dependent rows in a single transaction and run `SET CONSTRAINTS REFERENCES DEFERRED;` at the beginning of the tx, it looks like that might do the trick?

http://www.postgresql.org/docs/9.5/static/sql-set-constraints.html

**Comment by snarfed** [comment]
damn. i got antsy and tried this with snarfed/metabase@6cd85e1 (diff below), but it doesn't work. same error message. not sure why. :/

```diff
--- a/src/metabase/cmd/load_from_h2.clj
+++ b/src/metabase/cmd/load_from_h2.clj
@@ -78,6 +78,7 @@
         h2-db    (kdb/create-db (db/jdbc-details {:type :h2, :db (str filename ";IFEXISTS=TRUE")}))] ; TODO - would be nice to add `ACCESS_MODE_DATA=r` but it doesn't work with `AUTO_SERVER=TRUE`
     (db/setup-db)
     (kdb/transaction
+     (k/exec-raw (db/db-connection) "SET CONSTRAINTS ALL DEFERRED")
      (doseq [e     entities
              :let  [objs (kdb/with-db h2-db
                            (k/select (k/database e h2-db)))]
```

**Comment by snarfed** [comment]
yay, the fix worked great! thank you! i had to migrate my H2 db to 0.18 before running `load-from-h2`, but otherwise it ran happily.

**Comment by D-side** [comment]
Hello!

I'm pretty sure I'm hitting the same problem running `load-from-h2` with Metabase 0.19.1:

```
Transfering 17 instances of RawColumn....BatchUpdateException:
 Message: Batch entry 1 INSERT INTO raw_column ( "column_type", "raw_table_id", "name", "fk_target_column_id", "updated_at", "details", "active", "id", "is_pk", "created_at" ) VALUES ( NULL, 6, 'run_id', 55, '2016-08-09 18:50:00.672000 +03:00:00', '{"base-type":"IntegerField"}', '1', 39, '1', '2016-08-03 16:17:20.285000 +03:00:00' ) was aborted.  Call getNextException to see the cause.
 SQLState: 23503
 Error Code: 0
PSQLException:
 Message: ERROR: insert or update on table "raw_column" violates foreign key constraint "fk_rawcolumn_fktarget_ref_rawcolumn"
  Подробности: Key (fk_target_column_id)=(55) is not present in table "raw_column".
 SQLState: 23503
 Error Code: 0
```

* CentOS 6
* OpenJDK 1.8.0
* Metabase 0.19.1 `jar`
* embedded H2 => PostgreSQL 9.5

Reopen? Or could this be something else?

**Comment by D-side** [comment]
Apparently fixed by #3308, will check when the new version is released.

**Comment by camsaul** [comment]
@D-side correct, #3308 fixed this and #3151; the fix will ship in the next release.
